keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2336 @@
+that "structured parameters" provides.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+One of the primary motivating examples for supporting partitionable devices
+with DRA is to enable the dynamic allocation of Multi-Instance GPUs (MIG) on
+NVIDIA GPU hardware. MIG devices are represented as fixed-size partitions of a
+full GPU that consume some of its capacity across multiple dimensions. These
+dimensions include things like number of available JPEG engines, number of
+available multiprocessors, and the physical placement of a given partition
+across a set of fixed-size memory slices.
+
+The most important thing to note from this is that partitioning a GPU
+into a set of MIG devices is not as simple as slicing it up into a set of
+fixed-size, non-oerlapping partitions. Instead, individual MIG devices are defined by the
+set of (possibly overlapping) capacities that they consume from a full GPU. In
+other words, the allocation of one MIG device may nullify the ability to allocate some
+subset of other possible MIG devices (but not all of them). A MIG device is considered
+allocatable so long as the full GPU has enough *unallocated* capacity to satisfy
+its capacity constraints across all dimensions. If any single dimension is
+unavailable, that MIG device cannot be allocated (but others might still be able to).
+
+With his motivating example in mind. We define the following goals and
+non-goals of this KEP.
+
+### Goals
+
+* Introduce the ability for "structured parameters" DRA to allocate both full
+  devices and fixed-size partitions of full devices (across multiple
+  dimensions). Both full devices and their valid set of partitions must be
+  explicitly listed in a ResourceSlice for the scheduler to consider them for
+  allocation.
+
+* Provide abstractions to keep the overall footprint of a ResourceSlice small,
+  such that all full devices (and their partitions) can be represented in a
+  concise way.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+* Allow a user to allocate arbitrarily-sized partitions of a device. In other
+  words, only those partitions explicitly defined by a ResourceSlice will be
+  allocatable by a user. Arbitrary slicing is out of scope for this KEP.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+The basic idea is the following:
+
+1. Introduce a new device type called `PartitionableDevice` which has the same
```

Comment:
@pohly @johnbelamaric @thockin 
Curious about the naming of this new device type. I know we are adding this feature for the purpose of enabling "partitionable devices", but very few of the devices we define are actual "partitionable" -- the vast majority are them are the partitions themselves. Should we just call this `ExtendableDevice` (or something similar) to highlight the "mixin" nature of these new devices, and then just talk about the fact that these devices also have the ability to reference devices from which they can consume capactiy.

Thoughts?

Reply:
Sounds reasonable to me. This in fact may become the more common type, since you don't actually need it to be "partitionable" to gain compactness advantage from the extensibility. Probably `ExtensibleDevice` is better grammar.

Reply:
I agree that `PartititionableDevice` is emphasizing that aspect too much.

I'll keep out of the `ExtendableDevice` vs. `ExtensibleDevice` discussion: as non-native speaker I don't have a strong preference.

But let me share one thought: neither the device nor the type can be extended, unless a new field is safe to ignore. How about `AdvancedDevice` as the next step after `BasicDevice`?

Reply:
How about `CompositeDevice`?

Reply:
`AdvancedDevice` provides no semantic context, but yes, is aligned with "better than Basic". `CompositeDevice` sounds like it *might* be what we have been calling "compound devices" - a "meta" device that is used to allocate a multiple underlying devices as a unit. I know you are referring to "composing" the mixins, rather than composing devices, but it's a little fuzzy. Of course, I do think we can build a "compound device driver" on top of this...but that's not really what this is.

We can keep toying with some names, I'm not thrilled with any of those choices so far.

Reply:
I've changed it to "CompositeDevice" for now just to move away from "PartitionableDevice" (which I thing may have been confusing some of the readers since very few devices were actually partitioned).